File: YrradiaSDL2/Game/FPSCounter/FPSCounter.cs
namespace YrradiaSDL2.Game.FPSCounter
{
    public class FPSCounter
    {
        private int m_fps = 0;
        private int m_framesCount = 0;
        private int m_ticksLastUpdate = 0;
        public void Update()
        {
            if (Environment.TickCount >= m_ticksLastUpdate + 1000)
            {
                m_fps = m_framesCount;
                m_framesCount = 0;
                m_ticksLastUpdate = Environment.TickCount;
            }
            m_framesCount++;
        }

        public void Render()
        {
            var canvasSize = _.canvasUtilities.GetCanvasSize();
            _.textRenderer.DrawString("FPS: " + m_fps, new (canvasSize.Width - 200, 10), Colors.wheat);
        }
    }
}

File: src/Core/FPSCounter/FPSCounter.cpp
/*
 * Copyright 2025 Andreas Åkerberg
 * This code is licensed under MIT license (see LICENSE for details)
 */

#include "FPSCounter.hpp"
#include "Core/Rendering/Text/TextRenderer.hpp"

namespace Forradia
{
    void FPSCounter::Update()
    {
        auto now{GetTicks()};

        if (now > m_ticksLastUpdate + k_oneSecondMillis)
        {
            m_fps = m_framesCount;
            m_framesCount = 0;
            m_ticksLastUpdate = now;
        }

        ++m_framesCount;
    }

    void FPSCounter::Render() const
    {
        auto fpsText{"FPS: " + std::to_string(m_fps)};

        _<TextRenderer>().DrawString(
            fpsText,
            k_position.x,
            k_position.y);
    }
}
